src/object.d

```diff
@@ -4881,3 +4881,99 @@ void __ArrayDtor(T)(T[] a)
     foreach_reverse (ref T e; a)
         e.__xdtor();
 }
+
+/***********************************************************
+ * Used by `__ArrayCast` to obtain a descriptive error message.
+ * It is a template so it can be used by `__ArrayCast` in -betterC
+ * builds.  It is separate from `__ArrayCast` to minimize code
+ * bloat.
+ * Params:
+ *  fromType = name of the type being cast from
+ *  fromSize = total size in bytes of the array being cast from
+ *  toType   = name of the type being cast o
+ *  toSize   = total size in bytes of the array being cast to
+ * Returns:
+ *  a descriptive error message describing the cast failure
+ */
+private void onArrayCastError()(string fromType, size_t fromSize, string toType, size_t toSize) @trusted
+{
+    import core.internal.string : unsignedToTempString;
+    import core.stdc.stdlib : alloca;
+
+    const(char)[][8] msgComponents =
+    [
+        "Cannot cast `"
+        , fromType
+        , "` to `"
+        , toType
+        , "`; an array of size "
+        , unsignedToTempString(fromSize)
+        , " does not align on an array of size "
+        , unsignedToTempString(toSize)
+    ];
+
+    // convert discontiguous `msgComponents` to contiguous string on the stack
+    size_t length = 0;
+    foreach(m ; msgComponents)
+        length += m.length;
+
+    auto msg = (cast(char*)alloca(length))[0 .. length];
+
+    size_t index = 0;
+    foreach(m ; msgComponents)
+        foreach(c; m)
+            msg[index++] = c;
+
+    // first argument must evaluate to `false` at compile-time to maintain memory safety in release builds
+    assert(false, msg);
+}
+
+/***********************************************************
+ * The compiler lowers expressions of `cast(TTo[])TFrom[]` to
+ * this implementation.
+ * Params:
+ *  from = the array to reinterpret-cast
+ * Returns:
+ *  `from` reinterpreted as `TTo[]`
+ */
+TTo[] __ArrayCast(TFrom, TTo)(TFrom[] from) @trusted
+{
+    const fromSize = from.length * TFrom.sizeof;
+    const toLength = fromSize / TTo.sizeof;
+
+    if ((fromSize % TTo.sizeof) != 0)
```

Comment:
Making sure that @Geod24's suggestion in https://github.com/dlang/druntime/pull/2264#discussion_r207113286 isn't lost:
```d
version (D_NoBoundsChecks) {}
else if ((fromSize % TTo.sizeof) != 0)
```

Reply:
I'm waiting to find out what happens with https://github.com/dlang/dmd/pull/8532.  I don't think it should hold up this PR, and would even be better if it were done in a followup PR.

Reply:
Fair.